### DISCUSSION TITLE
Main thread hang detection

### DISCUSSION THREAD

**Discussion by jim-king-2000** [libuv/libuv#4203] [General]
Since it is impossible to write a perfect program, it would hang in some corner case even `libuv` is introduced into the system. We'd like to implement a mechanism to detect the hang of the main thread (maybe the event loop is blocked) and execute restart. Is there any best practices or suggestions here?

**Comment by bnoordhuis** [comment]
Watchdog functionality? Several ways to skin that cat but one way is to:

1. start a watchdog thread that blocks in uv_cond_timedwait()
2. if the timeout expires, the event loop thread is hung
3. uv_timer_start() a repeating timer in the event loop thread that calls uv_cond_signal()

**Comment by jim-king-2000** [comment]
Sounds great! Since it is indispensable, will `libuv` consider to implement it?

**Reply by bnoordhuis** [reply]
No. Libuv just provides you with the building blocks. Mechanism over policy.

**Reply by jim-king-2000** [reply]
OK. We'll try to implement it. Seems not so complex. At last, good wishes to this great library, it saves us from lock hell.
